subsys/net/lib/latmon/latmon.c

```diff
@@ -0,0 +1,563 @@
+{
+	int latmus = *(int *)p1;
+	struct latmon_data sample;
+
+	LOG_INF("Transfer thread priority: %d", XFER_THREAD_PRIORITY);
+
+	for (;;) {
+		if (k_msgq_get(&xfer_msgq, &sample, K_FOREVER)) {
+			LOG_ERR("Failed to get sample data to transfer");
+			continue;
+		}
+
+		if (send_sample_data(latmus, &sample) < 0) {
+			LOG_ERR("Failed to transfer sample data");
+			break;
+		}
+	}
+}
+
+static void start_xfer_thread(int *latmus)
+{
+	k_thread_create(&xfer_thread, xfer_thread_stack, XFER_THREAD_STACK_SIZE,
+			xfer_thread_func, latmus, NULL, NULL,
+			XFER_THREAD_PRIORITY, 0, K_MSEC(10));
+}
+
+static void abort_xfer_thread(void)
+{
+	k_thread_abort(&xfer_thread);
+}
+
+static int measure(uint32_t *delta, struct latmon_message *msg,
+		struct latmon_data *data,
+		struct latmus_conf *conf)
+{
+	if (data->warmed) {
+		k_usleep(conf->period);
+	}
+
+	if (msg->measure_func(delta) < 0) {
+		if (data->overruns++ > conf->max_samples / 2) {
+			return -1;
+		}
+		/* Just an overrun */
+		return 1;
+	}
+	return 0;
+}
+
+static void monitor_thread_func(void *p1, void *p2, void *p3)
+{
+	struct latmon_message *msg = p1;
+	struct latmus_conf *conf = p2;
+	struct latmon_data *data = p3;
+	uint32_t delta = 0;
+	int ret = 0;
+
+	LOG_INF("Monitor thread priority: %d", MONITOR_THREAD_PRIORITY);
+
+	/* Prepare transfer thread */
+	start_xfer_thread(&msg->latmus);
+
+	LOG_INF("\tmonitoring started:");
+	LOG_INF("\t - samples per period: %u", conf->max_samples);
+	LOG_INF("\t - period: %u usecs", conf->period);
+	LOG_INF("\t - histogram cells: %u", conf->cells);
+
+	/* Sampling loop */
+	memset(data, 0, sizeof(*data));
+	data->min_lat = UINT32_MAX;
+	data->warmed = false;
+	delta = 0;
+	while (!abort_monitor) {
+		for (;;) {
+			ret = measure(&delta, msg, data, conf);
+			if (ret) {
+				if (ret < 0) {
+					LOG_ERR("\tExcessive overruns, abort!");
+					goto out;
+				}
+				continue;
+			}
+
+			if (!prepare_sample_data(delta, conf, data)) {
+				/* Sample data ready */
+				break;
+			}
+		}
+
+		if (enqueue_sample_data(data) < 0) {
+			break;
+		}
+	};
+out:
+	abort_xfer_thread();
+	k_sem_give(&monitor_done);
+	monitor_tid = NULL;
+
+	LOG_INF("\tmonitoring stopped");
+}
+
+static int broadcast_ip_address(struct in_addr *ip_addr)
+{
+	char ip_str[NET_IPV4_ADDR_LEN];
+	struct sockaddr_in broadcast;
+	int sock = -1;
+	int ret = -1;
+
+	sock = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+	if (sock < 0) {
+		LOG_ERR("failed to create broadcast socket : %d", errno);
+		return -1;
+	}
+
+	broadcast.sin_addr.s_addr = htonl(INADDR_BROADCAST);
+	broadcast.sin_port = htons(LATMON_NET_PORT);
+	broadcast.sin_family = AF_INET;
+
+	net_addr_ntop(AF_INET, ip_addr, ip_str, sizeof(ip_str));
+
+	ret = zsock_sendto(sock, ip_str, strlen(ip_str), 0,
+			(struct sockaddr *)&broadcast, sizeof(broadcast));
+
+	zsock_close(sock);
+
+	return ret;
+}
+
+/* Get a socket to listen to Latmus requests */
+int net_latmon_get_socket(struct sockaddr_in *connection_addr)
```

Comment:
But here if we have `struct sockaddr *connection_addr`, then we do not need to tweak the API to support IPv6.
Then the memcpy() below could look like this
```
if (connection_addr != NULL) {
	memcpy(&addr, (struct sockaddr_in *)connection_addr, sizeof(addr));
}
```